File: data_structures/python/graph.py
# -*- coding: utf-8 -*-

"""
A proper, full function graph implementation.
Implemented as references.
"""

from typing import Dict, List

__author__ = "Jakrin Juangbhanich"
__email__ = "juangbhanich.k@gmail.com"


class Vertex:
    def __init__(self, key: str, data=None):
        self.key = key
        self.data = data
        self.edges: Dict[key, float] = {}


class Graph:
    def __init__(self, directed: bool=True, default_weight: float=1):
        self._default_directed: bool = directed
        self._default_weight: float = default_weight
        self._vertex_map: Dict[str, Vertex] = {}
        self._n_edges = 0

    def add_vertex(self, key: str, data=None):
        """ Add a vertex to this graph. """
        if key in self._vertex_map:
            self._vertex_map[key].data = data
        else:
            v = Vertex(key, data)
            self._vertex_map[key] = v

    def get_vertex(self, key: str):
        if key in self._vertex_map:
            return self._vertex_map[key]
        else:
            return None

    def __getitem__(self, key):
        return self.get_vertex(key).data

    def __setitem__(self, key, data):
        return self.add_vertex(key, data)

    def remove_vertex(self, key):
        """ Remove a vertex from this graph. All connected edges will also be removed. """
        for k, v in self._vertex_map.items():
            if key in v.edges:
                v.edges.pop(key)
                self._n_edges -= 1

        self._n_edges -= len(self._vertex_map[key].edges)
        self._vertex_map.pop(key)

    def add_edge(self, key_from, key_to, weight: float=None, directed: bool=None):
        """ Add an edge to this graph. If the vertex is not in, it will create one. """

        if directed is None:
            directed = self._default_directed

        if weight is None:
            weight = self._default_weight

        if key_from not in self._vertex_map:
            self.add_vertex(key_from, None)

        if key_to not in self._vertex_map:
            self.add_vertex(key_to, None)

        v_from = self._vertex_map[key_from]
        if key_to not in v_from.edges:
            self._n_edges += 1
        v_from.edges[key_to] = weight

        if not directed:
            v_to = self._vertex_map[key_to]
            if key_from not in v_to.edges:
                self._n_edges += 1
            v_to.edges[key_from] = weight

    def remove_edge(self, key_from, key_to, directed: bool=None):
        """ Remove an edge from the graph. """
        if directed is None:
            directed = self._default_directed

        if key_from not in self._vertex_map:
            raise Exception

        v_from = self._vertex_map[key_from]
        v_from.edges.pop(key_to)

        if not directed:
            v_to = self._vertex_map[key_to]
            if key_from in v_to.edges:
                v_to.edges.pop(key_from)
                self._n_edges -= 1

        self._n_edges -= 1

    def is_adjacent(self, key1: str, key2: str):
        return key2 in self._vertex_map[key1].edges

    def get_adjacent(self, key: str) -> List[Vertex]:
        """ Get a list of all vertices reachable by the given key. """
        n_keys = self._vertex_map[key].edges.keys()
        vertices = [self._vertex_map[k] for k in n_keys]
        return vertices

    def get_adjacent_keys(self, key: str) -> List[str]:
        """ Get the keys of all the reachable vertices from the given vertex. """
        return [v.key for v in self.get_adjacent(key)]

    @property
    def vertex_count(self) -> int:
        """ Number of vertices in the graph. """
        return len(self._vertex_map)

    @property
    def edge_count(self) -> int:
        """ All edges in the graph. Undirected edges will count as two edges. """
        return self._n_edges

    def display(self):
        """ Print out the graph into the std::out. """
        print("\n--- GRAPH ---")
        print(f"\nVERTS: {self.vertex_count}")
        if self.vertex_count > 0:
            print(" ".join(self._vertex_map.keys()))

        print(f"\nEDGES: {self.edge_count }")
        if self.edge_count > 0:

            seen_map = {}  # Prevent undirected arrows being printed twice.

            for k, v in self._vertex_map.items():
                for k2, w in v.edges.items():

                    if k in seen_map and seen_map[k] == k2:
                        continue

                    v2 = self._vertex_map[k2]
                    if k in v2.edges and v2.edges[k] == w:
                        print(f"{k} <--{w}--> {k2}")
                        seen_map[k2] = k
                    else:
                        print(f"{k} ---{w}--> {k2}")

        print("")

    # ===================================================================================================
    # Iterators.
    # ===================================================================================================

    def vertices(self):
        return self._vertex_map.items()

    def __iter__(self):
        return iter(self.vertices())

File: data_structures/go/graph.go
package ds

import "fmt"

type Vertex struct {
	Key   string
	Data  interface{}
	Edges map[string]float64
}

type Graph struct {
	Directed      bool
	DefaultWeight float64
	vertexMap     map[string]*Vertex
	numberOfEdges int
}

func NewGraph(directed bool) *Graph {
	graph := Graph{
		Directed:      directed,
		DefaultWeight: 1,
		vertexMap:     make(map[string]*Vertex, 0),
	}
	return &graph
}

func NewUndirectedGraph() *Graph {
	graph := NewGraph(false)
	return graph
}

func NewDirectedGraph() *Graph {
	graph := NewGraph(true)
	return graph
}

func (g *Graph) AddVertex(key string) {
	g.AddVertexData(key, nil)
}

func (g *Graph) AddVertexData(key string, data interface{}) {
	if v, ok := g.vertexMap[key]; ok {
		v.Data = data
	} else {
		vertex := Vertex{Key: key, Data: data, Edges: make(map[string]float64, 0)}
		g.vertexMap[key] = &vertex
	}
}

func (g *Graph) GetVertex(key string) *Vertex {
	// TODO: Handle Errors.
	return g.vertexMap[key]
}

func (g *Graph) GetData(key string) interface{} {
	// TODO: Handle Errors.
	return g.vertexMap[key].Data
}

func (g *Graph) RemoveVertex(key string) {

	// Reduce edges count.
	g.numberOfEdges -= len(g.vertexMap[key].Edges)
	delete(g.vertexMap, key)

	// Remove all associated edges.
	for _, v := range g.vertexMap {
		if _, ok := v.Edges[key]; ok {
			delete(v.Edges, key)
			g.numberOfEdges--
		}
	}
}

func (g *Graph) ensureVertex(key string) {
	if _, ok := g.vertexMap[key]; !ok {
		g.AddVertex(key)
	}
}

func (g *Graph) addEdgeWithWeightActual(keyFrom string, keyTo string, weight float64, reverse bool) {

	g.ensureVertex(keyFrom)
	g.ensureVertex(keyTo)

	vertex := g.vertexMap[keyFrom]

	// Increment Edge Count.
	if _, ok := vertex.Edges[keyTo]; !ok {
		g.numberOfEdges++
	}

	// Set the Edge.
	vertex.Edges[keyTo] = weight

	// If not directed, also add/modify the mirror.
	if reverse {
		g.addEdgeWithWeightActual(keyTo, keyFrom, weight, false)
	}
}

func (g *Graph) AddEdgeWithWeight(keyFrom string, keyTo string, weight float64) {
	g.addEdgeWithWeightActual(keyFrom, keyTo, weight, !g.Directed)
}

func (g *Graph) AddEdge(keyFrom string, keyTo string) {
	g.AddEdgeWithWeight(keyFrom, keyTo, g.DefaultWeight)
}

func (g *Graph) removeEdgePrivate(keyFrom string, keyTo string, reverse bool) {
	vertex := g.vertexMap[keyFrom]

	if _, ok := vertex.Edges[keyTo]; ok {
		delete(vertex.Edges, keyTo)
		g.numberOfEdges--
	}

	// If not directed, also add/modify the mirror.
	if reverse {
		g.removeEdgePrivate(keyTo, keyFrom, false)
	}
}

func (g *Graph) RemoveEdge(keyFrom string, keyTo string) {
	g.removeEdgePrivate(keyFrom, keyTo, !g.Directed)
}

func (g *Graph) IsAdjacent(keyFrom string, keyTo string) bool {
	vertex := g.vertexMap[keyFrom]
	if _, ok := vertex.Edges[keyTo]; ok {
		return true
	}
	return false
}

func (g *Graph) GetEdgeWeight(keyFrom string, keyTo string) float64 {
	vertex := g.vertexMap[keyFrom]
	if w, ok := vertex.Edges[keyTo]; ok {
		return w
	}
	return 0
}

func (g *Graph) GetAdjacent(key string) map[string]float64 {
	vertex := g.vertexMap[key]
	return vertex.Edges
}

func (g *Graph) GetAdjacentKeys(key string) []string {
	vertex := g.vertexMap[key]
	keyList := make([]string, 0)

	for k, _ := range vertex.Edges {
		keyList = append(keyList, k)
	}
	return keyList
}

func (g *Graph) VertexCount() int {
	return len(g.vertexMap)
}

func (g *Graph) EdgeCount() int {
	if g.Directed {
		return g.numberOfEdges
	} else {
		return g.numberOfEdges / 2
	}
}

func (g *Graph) Display() {
	fmt.Println("\n--- GRAPH ---")
	fmt.Printf("\n--- VERTS: %v ---\n", g.VertexCount())

	if g.VertexCount() > 0 {
		i := 0
		for k, _ := range g.vertexMap {
			fmt.Printf("%v", k)
			if i < len(g.vertexMap)-1 {
				fmt.Print(" ")
			}
		}
		fmt.Print("\n")
	}

	fmt.Printf("\n--- EDGES: %v ---\n", g.EdgeCount())
	if g.EdgeCount() > 0 {
		seenMap := make(map[string]string, 0)
		for k, v := range g.vertexMap {
			for k2, w := range v.Edges {

				if linkedKey, ok := seenMap[k]; ok && linkedKey == k2 {
					continue
				}

				seenMap[k2] = k

				if !g.Directed {
					fmt.Printf("%v <--%v--> %v\n", k, w, k2)
				} else {
					fmt.Printf("%v ---%v--> %v\n", k, w, k2)
				}
			}
		}
		fmt.Print("\n")
	}
}
